src/EditorFeatures/Core/Shared/Tagging/Utilities/TagSpanIntervalTree.TagNode.cs

```diff
@@ -2,49 +2,36 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
 using Microsoft.VisualStudio.Text;
 using Microsoft.VisualStudio.Text.Tagging;
 
 namespace Microsoft.CodeAnalysis.Editor.Shared.Tagging;
 
 internal partial class TagSpanIntervalTree<TTag>
 {
-    private class TagNode(ITagSpan<TTag> ts, SpanTrackingMode trackingMode)
+    private readonly struct TagNode(ITagSpan<TTag> tagSpan, SpanTrackingMode trackingMode)
     {
-        public readonly TTag Tag = ts.Tag;
-        public readonly ITrackingSpan Span = ts.Span.CreateTrackingSpan(trackingMode);
-        private SnapshotSpan _snapshotSpan = ts.Span;
+        private readonly ITagSpan<TTag> _originalTagSpan = tagSpan;
+        private readonly SpanTrackingMode _trackingMode = trackingMode;
 
-        private SnapshotSpan GetSnapshotSpan(ITextSnapshot textSnapshot)
-        {
-            var localSpan = _snapshotSpan;
-            if (localSpan.Snapshot == textSnapshot)
-            {
-                return localSpan;
-            }
-            else if (localSpan.Snapshot != null)
-            {
-                _snapshotSpan = default;
-            }
-
-            return default;
-        }
+        public TTag Tag => _originalTagSpan.Tag;
 
-        internal int GetStart(ITextSnapshot textSnapshot)
-        {
-            var localSpan = this.GetSnapshotSpan(textSnapshot);
-            return localSpan.Snapshot == textSnapshot
-                ? localSpan.Start
-                : this.Span.GetStartPoint(textSnapshot);
-        }
+        public TagSpan<TTag> GetTranslatedTagSpan(ITextSnapshot textSnapshot)
+            => new(GetTranslatedSpan(textSnapshot), Tag);
 
-        internal int GetLength(ITextSnapshot textSnapshot)
+        public SnapshotSpan GetTranslatedSpan(ITextSnapshot textSnapshot)
         {
-            var localSpan = this.GetSnapshotSpan(textSnapshot);
+            var localSpan = _originalTagSpan.Span;
+
             return localSpan.Snapshot == textSnapshot
-                ? localSpan.Length
-                : this.Span.GetSpan(textSnapshot).Length;
+                ? localSpan
+                : localSpan.TranslateTo(textSnapshot, _trackingMode);
         }
+
+        public int GetStart(ITextSnapshot textSnapshot)
```

Comment:
>GetStart [](http://example.com/codeflow?start=19&length=8)

Could this just be a GetSpan method instead, encouraging callers to not need to call both of these (which might help in the case where a translation is necessary)

Reply:
Duh, that's what GetTranslatedSpan is. I guess the question is why even have GetStart/GetLength

Reply:
the way interval trees work, we have a lot of algorithms that don't need the full span.  that said, i'll clean this up a bit.

Reply:
Even if they don't need the full span, isn't calling GetTranslatedSpan the same cost as calling GetStart or GetLength individually? 

The thought I had was if the GetStart/GetLength methods were removed, then callers that need one will just call the GetTranslatedSpan and use the appropriate Length/Start. Callers that need both can be trivially changed to just call GetTranslatedSpan once and use that result to get both the length and start.

Reply:
i think we need to chat.

Reply:
Note: my primary goals in all of these are not to make large scale refactorings to the components, but to address the stuff that profiles are calling out as a problem.

Reply:
Yeah, I get that. This is definitely better, and I'm going to go ahead and approve. 

Maybe I'm a bit confused on what this object is doing, I might take a peek once your changes are in.